Disabled cursor mode doesn't work right when connected over RDP #1276

Hilderin opened the pull request:
This is a new PR to fix the problem over Remote Desktop when Cursor Mode is Disabled.

https://github.com/glfw/glfw/issues/1276

Hilderin commented:
Sorry, i did not see your reply, I share your concern about this change when not using remote desktop. I'll try what you suggest soon and come back to you.

Hilderin commented:
I added an check if the current session in from remote desktop based on this article I found: https://learn.microsoft.com/en-us/windows/win32/termserv/detecting-the-terminal-services-environment
If not Remote Desktop, the original code should now execute and blankCursor is NULL has before.

Tested via Remote Desktop and locally on Windows 11 with the cursor exemple project.

dougbinks commented:
Many thanks, a first glance that looks great. I'll test and review shortly.

dougbinks commented:
Looking through the code I am think it would be more suitable if the cursor and isRemoteSession variable were in the  `_glfw.win32` variable (part of the `_GLFWlibraryWin32` struct) and initialized at startup. We only need one cursor for all windows, and hopefully only need to check for RDP at app startup. What are your thoughts?

Hilderin commented:
Excellent suggestion, i did the modification to move the isRemoteSession and blankCursor in _GLFWlibraryWin32 and i moved the code to init them in win_init. I really new in the glfw code, so i'm not sure i moved everything at the right place. Just tell me if everything sounds good.

dougbinks commented:
Looking at the [terms of use for learn.microsoft.com](https://learn.microsoft.com/en-us/legal/termsofuse) it seems that we likely can't use their code directly as there is no suitable license for the code.

However, luckily [RemoteFX](https://en.wikipedia.org/wiki/RemoteFX#RemoteFX_vGPU_Deprecation) has been deprecated and removed from all versions of windows, so we only need to use `GetSystemMetrics(SM_REMOTESESSION)` inline to detect a remote session.

In order to ensure no copyrighted code is in GLFW repository I can squash and merge this PR, so we just need to update to using the simpler check.

In completing this PR we will also need the steps as [Contributing a bug fix](https://github.com/glfw/glfw/blob/master/docs/CONTRIBUTING.md#contributing-a-bug-fix) namely:

 - Change log entry in README.md, describing the incorrect behavior
 - Credits entries for all authors of the bug fix

However if I can add these after merging.

Note that as yet I've not gone through rigorous testing and checking of the code, which I'll do once we have the RemoteFX change done.

Hilderin commented:
The code for RemoteFX has been removed and the Readme.md updated. But i don't known where and how to set "Credits entries for all authors of the bug fix".

dougbinks commented:
Thanks once again for this!

The credits are in [CONTRIBUTORS.md](https://github.com/glfw/glfw/blob/master/CONTRIBUTORS.md), sorted alphabetically by last name if using a full name or by handle if using that. I'll modify the guide to make that clear. I can also add your credit after merging so don't worry about this (as your handle Hilderin or your name if you want).

Could you remove the function `isCurrentRemoteSession()` and simply use `GetSystemMetrics(SM_REMOTESESSION) > 0`? That would simplify things and remove any trace of the article code. Although I could do this it might cause the merge to label the squashed commit as mine, and you've done all the work so I'd prefer to retain you as author.

Hilderin commented:
The function isCurrentRemoteSession has been removed.
Thanks to add Hilderin to the contributors for me.
I hope you have good vacation!
I'm here if you find problems later in your tests.

dougbinks commented:
Having tested this over Remote Desktop I can confirm the solution works.

However the issue also appears to be present with a virtual machine without a remote desktop connection, so I am investigating why (possibly related to RemoteFX as this is an older VM, so I will check if updating resolves the issue).

Hilderin commented:
Interesting, i did not test on a virtual machine. From what i read, the absolute mode should be used only on remote desktop. But with my tests, it is also used with other remote tools like Parsec. Maybe it is used for Virtual machine also and we should remove the isRemoteDesktop condition completely?

dougbinks commented:
I've tried setting `isRemoteDesktop` to `TRUE` and this doesn't resolve the VM issue, which I now believe to be a separate problem. So we should keep this test.

dougbinks commented:
I note that much of this code appears to be a copy-paste of #1279 by @Pokechu22 ?

Hilderin commented:
Yes absolutely, i based my modifications on this PR, i fixed the merge problem and made some minor modifications. Maybe i should have reference this PR from 2018 somewhere, i'm pretty new to all this. Feel free to include @Pokechu22 in the contributors.

Pokechu22 commented:
I'm fine with being added to the contributors without any other clarification (a link to my PR in the commit message or a [co-authored-by](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) tag would be nice, but if it ends up being difficult to add one then it doesn't really matter). I meant to rebase that PR a long time ago but never really got around to it (since I don't have a good setup for testing RDP issues anymore). Thanks for handling that for me :)

dougbinks commented:
Many thanks all - I'll try to add a co-authored-by tag for @Pokechu22  and will definitely add them to the GLFW contributors.

dougbinks commented:
I've submitted some change requests in my review. Once those are done I should be able to merge this PR, and then I will add the credits if you don't do so yourself.

Hilderin commented on src/win32_window.c:
```diff
@@ -925,9 +931,38 @@ static LRESULT CALLBACK windowProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM l
+                    // One other unfortunate thing is that re-centering the cursor will still fire an
+                    // input event; assume that any motion to the center is our re-centering and ignore it
+                    if (pos.x == window_width / 2 && pos.y == window_height / 2)
+                        break;
```
I think the point is to prevent raising twice the cursorPos callbacks when the cursor is recentered by code (in _glfwPollEventsWin32). I tested it over Remote Desktop and i hit the break on line 954.
![image](https://github.com/glfw/glfw/assets/81109165/f0700bd4-789c-40f5-9127-5f3f32285662)

You are right that the resulting is dx and dy are zero. If it's ok to raise the event twice when the mouse is moved, i can remove the check, but it's seems to me not a good idea? Maybe there's something else i don't undertand?

Hilderin replied:
Oh nevermind, after further check , I found a check in _glfwInputCursorPos that prevent raising cursorPos if the last virtualcursorX and Y and identical. That shoud prevent the cursorPos trice.

Hilderin commented:
I just pushed the requested modifications. I added @Pokechu22 and myself in the Contributors. I let you add the co-authored-by in the squash commit.

dougbinks commented:
Many thanks for making these changes. I will re-test later on today  and then ask @elmindreda for a final review.

dougbinks commented:
I've tested this updated version and as expected it works, enabling both disabled cursor and disabled cursor with raw mouse motion modes to work over a remote desktop connection.

dougbinks commented:
Merged, many thanks to you both. I was able to get @Pokechu22 as a co-author in the commit and @hilderin has already added them to the `CONTRIBUTORS.md` file.

Pokechu22 commented:
This fixes an issue I ran into with windows on both sides, and the underlying issue is a quirk with how RDP works on Windows. Whether or not it will affect Linux depends on a bunch of implementation details. It seems plausible that it could fix a similar issue with Linux but it depends on how e.g. xvnc behaves when the cursor is disabled. But it's worth trying at least.

Emeraldiaa commented:
okay, just one more question. do both devices need this? or only the host machine

Pokechu22 commented:
Only the game/other program you're running needs it.

dougbinks commented:
@Emeraldiaa what problem are you having / trying to solve?
